### PR TITLE
Implement In memory kv store in go 

### PR DESCRIPTION
**Pull request by s1ntaxe770r**

### CODE REVIEW

File: go/all-in-memory/pkg/kv/kv.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+func NewKv() *KVStore{
+	return &KVStore{
+		store: map[string]string{},
+	}
+}
+
+// Get returns the value of k  
+func ( kv *KVStore) Get(k string) string{
+	defer kv.mu.Unlock()
+	kv.mu.Lock()
```
Comment (Klaven) [inline go/all-in-memory/pkg/kv/kv.go]: should this not be `kv.mu.RLock()` As it stands you are locking it for all threads. This is not needed as you can have multiple readers at the same time on a map. Just not multiple writers.

**Review by Klaven — COMMENTED** [review]
two small comments.
1) (see in code)
2) No way to delete a value? only insert and read? - this becomes harder with things like replication and sharding so I think its important, even at this state when it's "easy"
3) I lied. three things. This looks good for the most part. Keep up the solid work. I love the readme, lets keep adding to the readme so if someone stubles across this repo it's golden without watching any of my vidoes :)

File: go/all-in-memory/pkg/kv/kv.go (hunk 1 above)
Comment (s1ntaxe770r) [reply go/all-in-memory/pkg/kv/kv.go]: Oh thanks, I was really confused as to which one to use

File: go/all-in-memory/pkg/kv/kv.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+func (kv *KVStore) Set(k, v string) string {
+	defer kv.mu.RUnlock()
+	kv.mu.RLock()
+	kv.store[k] = v
+	return k
+}
+
+func (kv *KVStore) Delete(k string) string {
+	defer kv.mu.RUnlock()
```
Comment (Klaven) [inline go/all-in-memory/pkg/kv/kv.go]: you are mutating here. this should be a "Lock", not rlock and runlock.

File: go/all-in-memory/pkg/kv/kv.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+// Get returns the value of k
+func (kv *KVStore) Get(k string) string {
+	defer kv.mu.RUnlock()
+	kv.mu.RLock()
+	val := kv.store[k]
+	return val
+}
+
+func (kv *KVStore) Set(k, v string) string {
+	defer kv.mu.RUnlock()
```
Comment (Klaven) [inline go/all-in-memory/pkg/kv/kv.go]: this is mutating. so it needs to be lock and unlock. not a reader lock.

**Review by Klaven — CHANGES_REQUESTED** [review]
I think I must have poorly explained myself. `RLock` is for reading, non-mutating code. aka `get` while the `Lock` is for mutating code and is mutually exclusive.

this looks to explain it well.
https://stackoverflow.com/questions/19148809/how-to-use-rwmutex

File: go/all-in-memory/pkg/kv/kv.go (hunk 2 above)
Comment (s1ntaxe770r) [reply go/all-in-memory/pkg/kv/kv.go]: Thank you for clarifying

**Review by Klaven — APPROVED** [review]
/lgtm

**Comment by s1ntaxe770r** [comment]
Shall I merge?

**Comment by Klaven** [comment]
can you rebase this and fit it into the "new" folder structure. I apologize. Originally I had your code in the re-organization but something happened as you know. I super apologize for the trouble.

**Pull request merged**